# 📝<font color=red>Bootcamp Solution</font>

## <font color=BLUE>A. Basic of Python</font>
###  Integer, Float, Boolean, String

In [3]:
print('Hello, World!')
print('Good Morning')
name = 'bidhan'
print(name)

Hello, World!
Good Morning
bidhan


In [4]:
# 1.
x = 3.5
print(float(int(x)))

3.0


In [5]:
# 2.
name = Varma
print(type(name))

NameError: name 'Varma' is not defined

In [6]:
# 3.
x = 3
y = 2.5
z = 5
print(X + y / z)

NameError: name 'X' is not defined

In [7]:
# 4.
4 % (1 + 9)**2 - 60 // (7 + 2)

-2

In [8]:
# 5.
a = (123)
b = '123'
c = 'Mike'
d = [111]

print(type(a))
print(type(b))
print(type(c))
print(type(d))

<class 'int'>
<class 'str'>
<class 'str'>
<class 'list'>


In [9]:
# 6. 
word = "I love Python programming"

print(word[7:13])
print(word[7:12])
print(word[-12:-18])
print(word[-18:-12])

Python
Pytho

Python


In [10]:
# 7. 
x = 'len'
x[len(x * 2) - 5]

'e'

#8.
String Problem

Description

In a certain encrypted message which has information about the location(area, city), the characters are jumbled such that first character of the first word is followed by the first character of the second word, then it is followed by second character of the first word and so on



In other words, let’s say the location is bandra,mumbai



The encrypted message says ‘bmaunmdbraai’.



Sample Input:

bmaunmdbraai



Sample Output:

bandra,mumbai



Let’s say the size or length of the two words wouldn’t match then the smaller word is appended with # and then encrypted in the above format.



With this in mind write a code to identify the right location and print it as place,city.

In [11]:
import ast,sys
input_str =input()
message1 = input_str[::2].strip('#')
message2 = input_str[1::2].strip('#')
print(message1, message2, sep=',')

bmaunmdbraai
bandra,mumbai


 or

In [12]:
import ast,sys
input_str = input()
message1 = input_str[0:-1:2]
message2 = input_str[1:len(input_str):2]
print(message1.strip('#') + "," + message2.strip('#'))


bmaunmdbraai#k
bandra,mumbaik


## Practice Question of Basic Python

In [13]:
# 9. Question 1: Let us start by printing the word "Welcome to the practice exercise" in python.

# Print the statement:
print('Welcome to the practice exercise')

Welcome to the practice exercise


In [14]:
# 10. Question 2: Create a string variable and print the string.

# Create a string variable:
string = " Hello, my dear friend !!!"
# Print the string:
print(string)

 Hello, my dear friend !!!


In [15]:
# 11. Question 3: Create a string by taking user input and print the string.
name = input('Enter ur name:')
# Print the string:
print(name)

Enter ur name:Bidhan Chandra Roy
Bidhan Chandra Roy


In [16]:
# 12. Question 4: Print the pattern given below: image.png

# Print the pattern shown in the question above:
print('  1',' 234','56789',sep='\n')

  1
 234
56789


In [17]:
# 13. Question 5: Print the '#' symbol 15 times without having to type it for 15 times.

# Print the pattern shown in the question above:
print('#'*15)

###############


In [18]:
# 14. Question 6: Take the user input to print the current date

date= input('enter the date:')

# Print the date:
date

enter the date:28-06-2022


'28-06-2022'

In [19]:
# 15. Question 7: Type cast the interger 1.8 into a string and print the result.

# Store the interger as a varaible:
x=1.8
# Typecast the integer:
y=str(x)
# Print the result:
y

'1.8'

In [20]:
# 16. Question 8: Find the product of two numbers input by the user.

# Take input of the first number and store it in a variable:
x=int(input("num1:"))
# Take input of the second number and store it in a variable:
y=int(input("num2:"))
# You will need to typecast the input strings into integers
# Calculate the product:
z=x*y
# Print the result:
print(z)

num1:25
num2:25
625


In [21]:
# 17. Question 9: Calculate the remainder of the number 6 when divided by the number 4.

# Find the modulus and print the result:
r=6%4
print(r)

2


In [22]:
# 18. Question 10: Solve the error in code below and print the correct result.

# Given:
# print('I can't study any longer')

# Corrected solution:
print('I can\'t study any longer')

I can't study any longer


In [23]:
# 19. Question 11: Print the following statement using only 1 print statement:

#Hello

#Good Morning

# Print the required statement:\
print('Hello\n\nGood Morning')


Hello

Good Morning


In [24]:
# 20. Question 12: Concatenate the given strings and print the result:

# Given below are the two strings:
string1= 'Welcome to upgrad'
string2= 'coding practice session.'

# Concatenate and print the result:
print(string1+' ' +string2)

Welcome to upgrad coding practice session.


In [25]:
#  21. Question 13: Find the length of string given below and print the result.

# Given below is the string:
string='Hi, I hope you are having a great learning experience.'
# Calculate and print the length:
len(string)

54

In [26]:
# 22. Question 14: Create a string of length=10. Obtain the fifth and seventh character of the string. Concatenate the characters and form a new string. Print the new string.

# Create a string of length 10:
string='abcdefghij'

# You can verify the length:
len(string)

# Obtain the fifth letter and store it into a new string:
string[4]

# Obtain the seventh letter and store it into a new string:
string[6]
# Concatenate the two strings:
a=string[4]+string[6]

# Print the result:
print(a)

eg


In [27]:
# 23. Question 15: Given below is a string. Perform the following operations and print the results:

# Remove the extra spaces from the string and print the new string.
# Find the length of new string
# Remove the last 5 characters of the new string
# Find the 15th character of the new string
# Compute the result of concatenation of the first and last letter of the new string.
# Given below is the string:

line= '     I hope you enjoyed this practice session.'

# Removing extra spaces:
s=line.lstrip(' ')
# Print the result:
s
# Calculate the length:
len(s)
# Slice the string to remove the last 5 characters:
p=s[:-5]
# Print the result:
p
# Print the 15th character of the string:
s[14]
# Print the concatenation of first and last character of resulting string:
q=p[0]+p[-1]
print(q)

Is


## Practice Excercise I

In [28]:
# 24. Consider the string given below and answer the following questions 
string = "OQYWFClFhFGAvIWYwGKpmZhnJiyzTslSIhSwvOsqJMEphzmifTkyqOMNpnOtXZxmCfgDYqbaBHAUvIWhMnvwZnEMVDvmEfLrDoQnAZgQEgXQVnmSYkfedpAdhrtpOgORpYLRZYGWdhWYuqQssCUXtTzKRDAhpjUheOzUroZNzWFtZOVwIapzUYtbSbjYNErzQ"
#Calculate the number of characters in the string

len(string)

193

In [29]:
# 25. Count the number of occurences of 'a' in the string

print(string.count('a'))

2


In [30]:
# 26. Identify whether the string starts with 'if'

print(string.startswith('if'))

False


In [31]:
# 27. Return the substring starting at index 63 and ending at index 88 [Both the indexes are inclusive]

print(string[63:89])

mCfgDYqbaBHAUvIWhMnvwZnEMV


In [32]:
# 28. Identify the character at index 45

print(string[45])

z


## <font color= blue>(Week-1) Bootcamp Practice Questions </font>
### <font color= blue> Practice Questions I </font>

In [33]:
# 28. What will be the output of the following code snippet?

number='159'

print(type(number))

<class 'str'>


In [34]:
# 29. What will be the output of the following code snippet?

number=34.721
print(float(int(number)))

34.0


In [35]:
# 30. What will be the output of the following code snippet?

name = Alex
print(type(name))

NameError: name 'Alex' is not defined

In [36]:
# 31. What will be the output of the following code snippet?

a = 8.9
b = 2
c = 5
print(a//b//c)

0.0


In [37]:
# 32. hat is the output of the following expression?

4 // 8 + ( 2.5 + 7.5 ) - 6 * 3

-8.0

In [38]:
# 33. What will be the output of the following expression?

5 % ( 2 + 12 ) ** 3 - 60 // ( 7 + 1 ) * 2

-9

In [39]:
# 34. What will be the output of the following code snippet?

string_1 = "I love to code in python"
print( string_1[7:13])

to cod


In [40]:
# 35. What is the code to extract “split” from the given string?

sentence = 'This sentence needs to be split into words'

print(sentence[-16:-11])
print(sentence[-17:-11])
print([sentence.split()[5]])


split
 split
['split']


In [41]:
# 36. What will be the output of the following code snippet?

len = 'python'
print(len(len))

TypeError: 'str' object is not callable

In [42]:
# 37. Identify the error in the following code fragment:

x=10
if x%2==0
print("even number")
else
print("odd number") 

SyntaxError: invalid syntax (3730134734.py, line 4)

In [43]:
# 38. What will be the output of the following code?
if 4+9<10:
    print("true")
else:
    print("false")
print("true")

false
true


In [44]:
# 39.  Consider the following code snippet: What will be the output if x=67, where x is the marks secured by a student?
x=67
if x>50:
    if x>90:
        print("excellent")
    else:
        if x>80:
            print("good")
        else:
            print("fair")
else:
    if x>30:
        print("average")
    else:
        print("poor")

fair


In [45]:
# 40. What will be output of the following code snippet?(no output)

a='pqrstuv'
i='j'
while i in a:
    print(i)

In [46]:
# 41. What will be the output of the following code snippet?(no output)

i = 16
while False:
    if i%2 == 0:
        break
    print(i)
    i += 2

In [47]:
# 42. What will be the output of the following code?

s="Upgrad"

for i in range(len(s)):
    if s[i] in 'a':
        s[i]='*'

print(s)

TypeError: 'str' object is not callable

### <font color= blue> Practice Questions II </font>

# 43.

__Write a program to display whether the input is a digit or a letter of the alphabet.__


In [48]:
# METHOD 1
test_char=input()
if test_char.isdigit():
    print('Integer')
elif test_char.isalpha():
    print('Aphabet')
else:
    print('Input is neither Alphabet nor Integer')


555
Integer


In [49]:
#  METHOD 2
ch=input()

if ((ch>='a' and ch<='z') or (ch>='A' and ch<='Z')):
    print('alphabet')
elif (ch>='0' and ch<='9'):
    print('Integer')
else:
    print('Input is neither Alphabet nor Integer')

bidhan
alphabet


# 44.

__Write a program to accept a character and display its next and previous character.__


__Hint: Make use of Ascii values here.__




In [50]:
character = input("Enter a character: ")

#get the Ascii value of the previous character using ord(ch)-1 
previous_character = chr(ord(character)-1)

#get the Ascii value of the next character using ord(ch)+1  
next_character = chr(ord(character)+1)

#print the characters

print(previous_character)
print(next_character)

Enter a character: m
l
n


# 45.

Write a program to accept a string from the user, delete all vowels from the string and display the result. 



In [51]:
# METHOD 1
string = input("enter the string: ")
vowels='aeiouAEIOU'
t=string.translate({ord(i): None for i in vowels})
print(t)

enter the string: Bidhan Chandra Roy
Bdhn Chndr Ry


In [52]:
# METHOD 2
# FOR LOOP TO ITERATE THE INPUT STRING
# CHARACTER WHICH ARE NOT VOWELS WILL BE ADDED TO NEW LIST

string = input('enter the string:')
vowels=['A','E','I','O','U','a','e','i','o','u']
output_string = ''
for i in string:
    if i not in vowels:
        output_string = output_string + i
print(output_string)

enter the string:What are you thinking dear?
Wht r y thnkng dr?


In [54]:
#METHOD 3
def vowel_removal(string):
    vowels= ['a','e','i','o','u']
    results=[character for character in string if character.lower() not in vowels]
    results= ''.join(results)
    print(results)

vowel_removal('Do not think too much') 

D nt thnk t mch


# 46. 

Write a program to accept three sides of a triangle as the input and print whether the triangle is valid or not.

(A triangle is valid if the sum of any two sides is greater than the third side.)


In [55]:
# METHOD 1

def triangle_validation(l1,l2,l3):
    if (l1+l2)>l3 and (l2+l3)>l1 and (l3+l1)>l2:
        return 'valid'
    else:
        return "Invalid"
    
triangle_validation(2,2,1)

'valid'

In [58]:
# METHOD 2

a,b,c= input().split()
a=int(a)
b=int(b)
c=int(c)

if (a+b)>c and (b+c)>a and (c+a)>b:
    print('Valid')
else:
    print('Invalid')


2 2 4
Invalid


### <font color= blue> Practice Questions III </font>

# 47.

Write a program to display the multiplication table of a given number.




In [59]:
# method 1

number= int(input('enter the number:'))
count=1

while count<11:
    print(f'{number} x {count} = {number*count}')
    count+=1

enter the number:5
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


In [60]:
# method 2

number= int(input('enter the number:'))

for i in range(1,11):
    print(f'{number} x {i} = {number*i}')

enter the number:3
3 x 1 = 3
3 x 2 = 6
3 x 3 = 9
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27
3 x 10 = 30


In [61]:
# method 3

number= int(input('enter the number:'))

[(f'{number} x {i} = {number*i}') for i in range(1,11)]


enter the number:25


['25 x 1 = 25',
 '25 x 2 = 50',
 '25 x 3 = 75',
 '25 x 4 = 100',
 '25 x 5 = 125',
 '25 x 6 = 150',
 '25 x 7 = 175',
 '25 x 8 = 200',
 '25 x 9 = 225',
 '25 x 10 = 250']

# 48.

Write a program to accept a string value from the user and accept a char value from the user and find out the total occurrence of the char value in the string value. Note that the count is not case-sensitive



In [62]:
# Method 1 
def char_count(string,char):
    x=string.lower()
    y=char.lower()
    count=x.count(y)
    return count
char_count('Bidhan CHANDRA Roy','a')

3

In [63]:
# Method 2

# Taking a user input string and a character
input_string=input()
input_char=input()

# Converting both the string and the chareacter in lower case to avoid the case sensitivity
input_string=input_string.lower()
input_char=input_char.lower()

# Printing total no of occurance of the character in the given string by using count() method
print(input_string.count(input_char))

'Bidhan CHANDRA Roy'
n
2


In [2]:
# Method 3

input_string=input().lower()
input_char=input().lower()
count=0

for i in range(len(input_string)):
    if (input_string[i]==input_char):
        count=count+1
print(count)



uff tired of tYPING
I
2


In [3]:
# Method 4

input_string=input().lower()
input_char=input().lower()
count=0
idx=0

while idx<len(input_string):
    if input_string[idx]==input_char:
        count=count+1
    idx+=1
print(count)
    

Getting bored
e
2


# 49.

Write a program to calculate the sum of the digits of a given number 



In [4]:
# Method 1

n=int(input())
n=str(n)
sum_digits=0

for digit in n:
    sum_digits=sum_digits+int(digit)

print(sum_digits)

255654
27


In [5]:
# Method 2

n=int(input())
n=str(n)
sum_digits=0
while n!=0 :
    sum_digits= sum_digits + (int(n)%10)
    n=int(n)/10
print(sum_digits)

253
10


In [6]:
# Method 3

def sum_of_digit(num):
    sum=0
    for digit in str(num):
        sum=sum+int(digit)
    return sum

sum_of_digit(555) 

15

# 50.

Write a program to accept a number from the user and count the zeros, odd digits and non-zero even digits from the entered number.



In [7]:
# Method 1
# taking input from the user
n=int(input())
#Creating the three variable to count the occurance of zero valued, even number and odd number and stroded in these three variable
n_zero=0
n_non_zero_even=0
n_odd=0

#Initialising a loop through the input string 
#where n_zero will count num of time zero occured inside the string and so for the n_non_zero_even and n_odd.
for num in str(n):
    if int(num)==0:
        n_zero+=1
        # Each occurance of zero will increase the count by 1 
    elif int(num)%2==0 and int(num)!=0:
        n_non_zero_even+=1
        # Each occurance of a non-zero which is even will increase the count by 1 
    elif int(num)%2!=0:
        n_odd+=1
        ## Each occurance of a non-zero which is odd will increase the count by 1 

# Printing the count of the three type of number
print(f'Number of odd digits: {n_odd}')
print(f'Number of non-zero even digits: {n_non_zero_even}')
print(f'Number of zeros: {n_zero}')

102546300014789965441233
Number of odd digits: 11
Number of non-zero even digits: 9
Number of zeros: 4


# 51.


Write a program that takes a string as the input and does the following:

    -Removes the numbers, special characters

    -Converts uppercase letters to lowercase letters, and vice versa 

   

In [1]:
# Method 1

# Taking a user input string 
input_string=input()

# By using swapcase() method swapping the upper case letters to lower and vice verca
input_string=input_string.swapcase()

# Creating an empty variable to take new string to store the swapped letters by removing number,special characters etc.
new_str= ''

# Initialising the loop through the input string to add only all the letters from the input string to new string
for alph in input_string:
    if alph.isalpha():
        new_str = new_str + alph

# Printing the new string after removing all the numbers, characters etc 
print(new_str)

bIDHAN chanDRA Roy @345681%%$%$$
BidhanCHANdrarOY


# <font color=BLUE>B. Data Structure in Python</font>

### List , Tuple, Set, Dictionary

In [10]:
# 1.
#Suppose list_1 is [2, 33, 222, 14, 25]. What is list_1[-5]?

list_1 = [2, 33, 222, 14, 25]
list_1[-5]

2

In [11]:
# 2. What would be the output of the code below?

word = ['1','2','3','4']
word[ : ] = [ ] 
print(word)

[]


In [12]:
# 3.

L = ['one','two','three', 'four', 'five', 'six']
print(sorted(L))
print (L)

['five', 'four', 'one', 'six', 'three', 'two']
['one', 'two', 'three', 'four', 'five', 'six']


# 4. 

List_remove_append

Description

Remove SPSS from input_list=['SAS', 'R', 'PYTHON', 'SPSS'] and add 'SPARK' in its place.

In [13]:
# Method 1

input_list=['SAS', 'R', 'PYTHON', 'SPSS']
input_list.remove('SPSS')
input_list.append('SPARK')
print(input_list)

['SAS', 'R', 'PYTHON', 'SPARK']


# 5.
List to String

Description

Convert a list ['Pythons syntax is easy to learn', 'Pythons syntax is very clear'] to a string using ‘&’. The sample output of this string will be:

Pythons syntax is easy to learn & Pythons syntax is very clear

Note that there is a space on both sides of '&' (as usual in English sentences).

In [14]:
# Method 1

#Input as mentioned in the problem.
input_list = ['Pythons syntax is easy to learn', 'Pythons syntax is very clear']

#By using insert() method , adding "&" in list by mentioning it's position i.e 1 and then following it's value '&'.
input_list.insert(1,'&')

#By using join() of the builtin function element of the list have been jioned and separateted by space.
list1=' '.join(input_list)

#Printing the new list 'list1'
print(list1)

Pythons syntax is easy to learn & Pythons syntax is very clear


In [15]:
# Medhod 2

#Input as mentioned in the problem.
input_list = ['Pythons syntax is easy to learn', 'Pythons syntax is very clear']

# Using append() , adding '&' inside the list and it will be added in the end as the last element of the list
input_list.append('&')

# creating a new string with the help of indexing
list1 = input_list[0] +' '+ input_list[2] + ' '+ input_list[1]

#Printing the new list 'list1'
print(list1)

Pythons syntax is easy to learn & Pythons syntax is very clear


In [16]:
# Method 3

#Input as mentioned in the problem.
input_list = ['Pythons syntax is easy to learn', 'Pythons syntax is very clear']

#By using join() of the builtin function element of the list have been jioned and separateted by 'space&space'.
list1=' & '.join(input_list)

#Printing the new list 'list1'
print(list1)


Pythons syntax is easy to learn & Pythons syntax is very clear


In [17]:
# 6.

#How will I extract Python from the nested list

input_list =  [['SAS','R'],['Tableau','SQL'],['Python','Java']]

print(input_list[2][0])


Python


# 7.

String Split

Description

Split the string input_str = 'Kumar_Ravi_003' to the person's second name, first name and unique customer code. In this example, second_name= 'Kumar', first_name= 'Ravi', customer_code = '003'.



A sample output of the input 'Kumar_Ravi_003' is:

Ravi
Kumar
003
 

Note that you need to print in the order first name, last name and customer code.

In [18]:
# Method 1

# Input as mentioned in the problem.
input_str = 'Kumar_Ravi_003'

# using split() method to remove '_' from the string and separate the desired substring and stored in a list.
str1 = input_str .split('_')

# Printing the desired output by string concatenation 
print(str1[1]+'\n'+str1[0]+'\n'+str1[2])

Ravi
Kumar
003


In [19]:
# Method 2

# Input as mentioned in the problem.
input_str = 'Kumar_Ravi_003'

# Using split() method to remove '_' from the string and separate the desired substring and stored in a list.
str1 = input_str .split('_')

# Creating variables for each substrings:
first_name = str1[1]
second_name = str1[0]
customer_code = str1[2]

# Printing the created variable in the desired sequence
print(first_name)
print(second_name)
print(customer_code)


Ravi
Kumar
003


In [20]:
# Method 3

# Input as mentioned in the problem.
input_str = 'Kumar_Ravi_003'

# Using split() method to remove '_' from the string and separate the desired substring and stored in a list and againing strong all the elements in different variables.
first_name = input_str.split("_")[1]
second_name = input_str.split("_")[0]
customer_code = input_str.split("_")[2]

# Printing the created variable in the desired sequence
print(first_name)
print(second_name)
print(customer_code)

Ravi
Kumar
003


In [21]:
# 8.

#What would be the output here?
t = ("disco", 12, 4.5)
t[0][2]

's'

# 9.
Suppose t = (1, 2, 4, 3). Which of the following statements would result in an error?


In [22]:
print(t[4])


IndexError: tuple index out of range

In [23]:
print(t[1:-4])


()


In [24]:
print(t[3] = 5)

SyntaxError: expression cannot contain assignment, perhaps you meant "=="? (1106023495.py, line 1)

In [25]:
print(len(t))

3


# 10.
Which of the following declarations can be used to create a Python tuple?


In [26]:
x = ('2')
type(x)

str

In [27]:
x = 1,2,
type(x)

tuple

In [28]:
x = ((1,2,3), 4,5)
type(x)

tuple

In [29]:
x = (Hello,'2','3')
type(x)

NameError: name 'Hello' is not defined

# 11.

Tuple

Description

Add the element ‘Python’ to a tuple input_tuple = ('Monty Python', 'British', 1969). Since tuples are immutable, one way to do this is to convert the tuple to a list, add the element, and convert it back to a tuple.

In [30]:
# Method 1

# Input as mentioned in the problem.
input_tuple = ('Monty Python', 'British', 1969)

# Converting tuple into list:
lst1=list(input_tuple)

# Adding 'Python' in the list 'lst1'
lst1.append('Python')

# Again converting the list into tuple to get the desired result
tuple_2=tuple(lst1)

# printing the desired output
print(tuple_2)

('Monty Python', 'British', 1969, 'Python')


# 12.
Sets Practice

Description

Let’s say you have two lists A and B. Identify the elements which are common in the two lists A and B and return them in a sorted manner. For example 



Sample Input :

A = [5, 1, 3, 4, 4, 5, 6, 7]

B = [3, 3, 5, 5, 1 ,7 ,2]



Sample Output:

[1,3,5,7]



If you observe the sample output here you can see that 

Though value 5 is repeated twice in both lists, in the final output it is present only once.
The values are returned in a sorted manner with values increasing.

In [31]:
# Method 1

# Input as mentioned in the problem.
A = [5, 1, 3, 4, 4, 5, 6, 7]
B = [3, 3, 5, 5, 1 ,7 ,2]

# converting lists into set to remove all the duplicates
setA = set(A)
setB = set(B)

# To get the common items present in both the lists , intersection(&) is used and then converted to a list
common_elements_lists =list(setA & setB)

# printing the desired output
print(common_elements_lists)

[1, 3, 5, 7]


In [32]:
# 13.
#What gets printed with the following set of instructions?

nums = set([1,1,2,3,3,3,4])
print(len(nums))

4


# 14.

Dictionary Coding

Description

Write code to fetch the profession of the employee with Employee id - 104 from an employee input given in the form of a dictionary where key represent employ id and values represent the name, age, and profession (in the same order).

Sample input:

Employee_data = { 101:['Shiva', 24, 'Content Strategist'] ,102:['Udit',25,'Content Strategist'], 103:['Sonam', 28,'Sr Manager'], 104:['Ansari',29,'Product Lead' ],105:['Huzefa',32,'Project Manager' ]}

Sample output:

'Product Lead'

Note: Assume that employee data would be available in the data provided to you 

In [33]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
input_dct= ast.literal_eval(input())

# As the employee id (eg: 104) is a key and the profession is in the 3rd place of dictionary value, we will call the profession by calling the key(employee id) and value of key using list indexing:
profession = input_dct[104][2]

# printing the profession of 104 :
print(profession)

{ 101:['Shiva', 24, 'Content Strategist'] ,102:['Udit',25,'Content Strategist'], 103:['Sonam', 28,'Sr Manager'], 104:['Ansari',29,'Product Lead' ],105:['Huzefa',32,'Project Manager' ]}
Product Lead


# 15.

Which of the following statements create a dictionary? Select all the options that apply. (More than one option may be correct.)

In [34]:
d = {}
type(d)

dict

In [35]:
d = {'a':1, 'b':2}
type(d)

dict

In [36]:
d = dict(a=1, b=2)
type(d)

dict

In [37]:
d = ('a':1, 'b':2)
type(d)

SyntaxError: invalid syntax (658412872.py, line 1)

In [38]:
# 16.
# Suppose dict_1 = {"Python'':40, "R'':45}. What command should be used to delete the entry "R"?
dict_1 = {"Python":40, "R":45}
del dict_1['R']
dict_1

{'Python': 40}

In [39]:
# 17.
#What would be the output of the following set of instructions?

d = {'Python':40, 'R':45}
print(list(d.keys()))

['Python', 'R']


# 18.
Dict_Error

Description

From a Dictionary input_dict={'Name': 'Monty', 'Profession': 'Singer' }, get the value of a key ‘Label’ which is not a part of the dictionary, in such a way that Python doesn't hit an error. If the key does not exist in the dictionary, Python should return 'NA'.


Sample Input:

{'Name': 'Monty', 'Profession': 'Singer' }


Sample Output:

NA

In [41]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
input_dct= ast.literal_eval(input())

# To search of looking for a value by using key get() method is used. and we can instruct the get() method that if the key is not present in the dictionary what will be our desired output:
output = input_dct.get('Label','NA')

# printing the output
print(output)

{'Name': 'Monty', 'Profession': 'Singer' }
NA


In [42]:
# Method 2

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will be  dictionary) of data structure as input literal_eval() method is used.
import ast

#taking dictionary as input
input_dct= ast.literal_eval(input())

# taking string as input to check weather it is present in the input dictionary
search_key = input()

keys = input_dct.keys()

if search_key in keys:
    print(input_dct[search_key])
else:
    print('NA')

{'Name': 'Monty', 'Profession': 'Singer' }
lABEL
NA


# 19. 

List of Values in a Dictionary.

Description

Create a SORTED list of all values from the dictionary input_dict = {'Jack Dorsey' : 'Twitter' , 'Tim Cook' : 'Apple','Jeff Bezos' : 'Amazon' ,'Mukesh Ambani' : 'RJIO'}

Sample Input:

{'Jack Dorsey' : 'Twitter' , 'Tim Cook' : 'Apple','Jeff Bezos' : 'Amazon' ,'Mukesh Ambani' : 'RJIO'}

Sample Output:

['Amazon', 'Apple', 'RJIO', 'Twitter']

In [43]:
# Method 1

# Input as mentioned in the problem.
input_dict = {'Jack Dorsey' : 'Twitter' , 'Tim Cook' : 'Apple','Jeff Bezos' : 'Amazon' ,'Mukesh Ambani' : 'RJIO'}

# Getting the list of values of the given dictionary
values_list = input_dict.values()

# To sort the value list followig the sorted() method and printing the output
output = sorted(values_list)

print(output)

['Amazon', 'Apple', 'RJIO', 'Twitter']


In [46]:
# 20.
#Question 1: Print a random integer between 100 and 135


# Import required libraries:
import random

# Print a random integer in the given range:
print(random.randint(100,135))

132


In [49]:
# 21.
#Question 2: Create a list consisting of first 10 even numbers.

# Create a list according to the given condition:
list=[0,2,4,6,8,10,12,14,16,18]
list

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

In [52]:
# 22.
#Question 3: Create a list in the following manner: [birthyear, firstname, birthdate, lastname]

# Create a list according to the given condition:
list1=[1994,'bidhan',25,'roy']
list1

[1994, 'bidhan', 25, 'roy']

In [53]:
# 23.
## Create a list according to the given condition:
list2=[['Alex','Cody'],['mathematics','english'],[97,95],[89,92]]
list2

[['Alex', 'Cody'], ['mathematics', 'english'], [97, 95], [89, 92]]

In [54]:
#24.
#Question 5: Given below is a list, find out whether is has the letter 'k' in it and print the result as true or false.

# Given below is a list:
list3=[1,2,'r',7,'k',6,'o','h','d',5,9]

# Define the character you are searching for:
x='k'

# print the result:
x in list3

True

In [55]:
#25.
# Question 6: In the tuple given below, print the 3rd element.

# Given below is the tuple:
tuple1=(1,'hello','world','567')

# Print the 3rd element:
tuple1[2]

'world'

In [56]:
# 26.
# Question 7: Given below is a tuple. Sort the tuple and print the result.

# Given below is the tuple:
tuple2=(1,25,67,100,3,54)

# Sort the tuple:
x= sorted(tuple2)
# Print the result:
x

[1, 3, 25, 54, 67, 100]

In [58]:
# 27.
# Question 8: Given below is a list, convert into a set and print the output also print the length of the set.

# Given below is the list:
list4=[1,2,4,6,3,2,5,4,7,4,6,5,3,2,2,1]

# Convert the list into set:
set1=set(list4)

# Print the result:
print(set1)

# Print the length of the set:
print(len(set1))

{1, 2, 3, 4, 5, 6, 7}
7


In [61]:
# 28.
# Question 9: Given below are two sets, perform the operations given below:

#Print their union
#Print their intersection
#Print their difference
#Print their symmetric difference



# Given below are two sets:
a={1,2,3,5,6,7}
b={4,5,8,9,2,3}

# Print their union:
union=a|b
print(union)
# Print their intersection:
intersection=a&b
print(intersection)
# Print their difference:
diff=a.difference(b)
print(diff)
# Print their symmetric difference:
symm=a^b
print(symm)

{1, 2, 3, 4, 5, 6, 7, 8, 9}
{2, 3, 5}
{1, 6, 7}
{1, 4, 6, 7, 8, 9}


In [64]:
# 29.
# Question 10: Create a dictionary that stores the names and roll numbers of 5 students as given below:

"""Sam, 21
Cody, 12
Addy, 3
Zach, 45
Amy, 6 """

# Create a dictionary:

dict1={'Sam':21,'Cody':12,'addy':3,'Zach':45,'amy':6}

# Print the result:
dict1

{'Sam': 21, 'Cody': 12, 'addy': 3, 'Zach': 45, 'amy': 6}

In [66]:
# 30.
#Question 11: Due to admission of new students the roll number of zach has changed from 45 to 47. Make the required changes and print the new dictionary.

# Update the roll number:
dict1['Zach']=47

# Print the new dictionary:
dict1

{'Sam': 21, 'Cody': 12, 'addy': 3, 'Zach': 47, 'amy': 6}

In [69]:
# 31.
# Question 12: Sort the dictionary according to roll number.

# Sort and print the students roll number wise in ascending order:

x=sorted(dict1)
x

['Cody', 'Sam', 'Zach', 'addy', 'amy']

In [89]:
# 32.
# List of order ID’s which are processed 
processed_orders = [1152, 1154, 1155, 1156, 1157, 1160, 1161, 1162, 1166, 1169, 1170, 1172, 1176, 1050, 1178, 1051, 1052, 1054, 1058, 1060, 1061, 1062, 1065, 1066, 1067, 1068, 1069, 1076, 1077, 1080, 1081, 1083, 1091, 1085, 1088, 1089, 1131, 1092, 1094, 1095, 1099, 1102, 1103, 1104, 1106, 1107, 1108, 1109, 1111, 1117, 1119, 1121, 1150, 1128, 1129, 1136, 1137, 1139, 1140, 1141, 1144, 1148, 1124]

# List of order ID’s which are returned
returned_orders = [1153, 1158, 1159, 1163, 1164, 1165, 1167, 1168, 1171, 1173, 1174, 1175, 1177, 1053, 1055, 1056, 1057, 1059, 1063, 1064, 1070, 1071, 1072, 1073, 1074, 1075, 1078, 1079, 1082, 1084, 1086, 1087, 1090, 1093, 1096, 1097, 1098, 1100, 1101, 1105, 1110, 1112, 1113, 1114, 1115, 1116, 1118, 1120, 1122, 1123, 1125, 1126, 1127, 1130, 1132, 1133, 1134, 1135, 1138, 1142, 1143, 1145, 1146, 1147, 1149, 1151]

# Consider the information available in the above two lists and answer the question given below
#Count the total number of orders [ Orders include both processed and returned orders]
total_order = processed_orders + returned_orders
print('total_order is ',len(total_order))

#In the total orders, identify the 50th order [ Note: Assume the order ID’s are being generated in a consecutive manner]
total_order.sort()
print('50th order is ',total_order[49])

#Is 50th order a returned order or processed order?
print(1099 in processed_orders)

#What is the last processed order ID ? [ Note: Assume the order ID’s are being generated in a consecutive manner]
'''
1050
1178
1124
1177'''
print(total_order[-1])

#Identify the first 4 orders which are processed?

processed_orders.sort()
print(processed_orders[0:4])



total_order is  129
50th order is  1099
True
1178
[1050, 1051, 1052, 1054]


In [115]:
# 33.
# employee data with key as id of the employee and values as age of the employee
Employee_data ={101: 43, 102: 25, 103: 43, 104: 31, 105: 26, 106: 28, 107: 29, 108: 43, 109: 25, 110: 22, 111: 22, 112: 25, 113: 30, 115: 45, 116: 23, 117: 29, 118: 28, 119: 30, 120: 28, 121: 42, 122: 39, 123: 29, 124: 42, 125: 43, 126: 42, 127: 40, 128: 27, 129: 23, 130: 30, 131: 37, 132: 20, 133: 36, 134: 27, 135: 27, 136: 22, 137: 28, 138: 23, 139: 45, 140: 39, 141: 29, 142: 33, 143: 39, 145: 34, 146: 26, 147: 30, 148: 38, 149: 29, 150: 24, 151: 28, 152: 34, 153: 42, 154: 29, 155: 23, 156: 31, 158: 25, 160: 45, 161: 42, 162: 27, 163: 24, 164: 20, 166: 24, 167: 28, 168: 20, 169: 33, 170: 34, 171: 37, 172: 45, 173: 35, 174: 23, 175: 44, 176: 27, 177: 30, 178: 26, 179: 27} 

#Identify the senior most employees age
x=sorted(Employee_data.values())
print(x[-1])

#Identify the age of the employee with employee id 159 [ If the employee isn't present return NA]¶
print(Employee_data.get(159,'Error'))

#Count the total number of employee in the organization
print(len(Employee_data))

#Calculate the mean age of the employees
mean_age= sum(Employee_data.values())/74
print(mean_age)

#Perform the following two tasks and then calculate the updated mean age of the employees
'''Task1 Update the ages of employee id - 104,140, and 164 as 27
Task2 Remove the employ with employee id - 143'''

Employee_data[104]=Employee_data[140]=Employee_data[164]=27

''' or we can do:
d1={104:27,140:27,164:27}
Employee_data.update(d1)
'''
del Employee_data[143]
mean_age= sum(Employee_data.values())/len(Employee_data)
print(mean_age)


45
Error
74
31.364864864864863
31.136986301369863


## <font color= blue>(Week-2) Bootcamp Practice Questions </font>
### <font color= blue> Practice Questions MCQ </font>

In [116]:
# 34.
#What is the output of the following?

s = 'a$b@c@d'
print(s.split('@','$'))

TypeError: 'str' object cannot be interpreted as an integer

In [117]:
# 35.
#What is the code to print the element ‘banana’?

fruits = ['apple', 'mango', ['grapes', 'papaya'], ('banana', 'pomegranate')]
fruits[3][0]

'banana'

In [118]:
# 36.
# Find the output of the following code.

list_1 = ['red' , 'orange', 'blue', 'green', 'black', 'white']
print(list_1.sort())

None


In [119]:
# 36.
# What will the output of the following code be?

num_1=[9,4,2,7]
num_2=[8,3,6,1]
print(sorted(num_1))
print(num_1)
print(num_2.sort())
print(num_2)

[2, 4, 7, 9]
[9, 4, 2, 7]
None
[1, 3, 6, 8]


In [120]:
# 37.
#Find the output of the following code.

l = ['paris','india','china','london']
print(tuple(l+['sri lanka']))

('paris', 'india', 'china', 'london', 'sri lanka')


In [121]:
# 38.
#What is the output of the following code?

var=(5)
print(type(var))

<class 'int'>


In [126]:
# 39.
#What is the output of the code given below?

tup=([1,3,5],"abcd",(2,4),6)
tup[0][2]=7
print(tup)

([1, 3, 7], 'abcd', (2, 4), 6)


In [127]:
# 40.
#What will the output of the following code be?

a= {1,2,4,5}
b= {5,6,7,8}
print(a-b)
 

{1, 2, 4}


In [128]:
# 41.
#What will the output of the following code be?

s= {3,4,5}
s*2

TypeError: unsupported operand type(s) for *: 'set' and 'int'

In [129]:
# 42.
#What is the output of the following code?

numbers= {1,0,6,1,4,0,7}
print(len(numbers))

5


In [130]:
# 43.
#What will the output of the following Python code be?

dict= {'ab':12,'cd':34,'ef':56}
print(dict[1])

KeyError: 1

In [131]:
# 44.
#What will the output of the following code be?  

marks= {'John':45,'Alex':60}
print(marks.keys())

dict_keys(['John', 'Alex'])


In [132]:
# 45.
#What is the output of the following code?

d = {"Python":40, "R":45}
"Python" in d

True

In [133]:
# 46.
#What will the output of the following code be?

input_dict= {1:"one",2:"two",3:"three"}
for i in input_dict:
    print(i) 

1
2
3


In [136]:
# 47.
#Find the output of the following code snippet.

input_str="I love programming in python"
count=0
l=['a','e','i','o','u']
for i in input_str.lower():
    if i in l:
        count=count+1
print(count)

8


### <font color= blue> Practice Questions I </font>

# 48.
List sum

Description

Suppose you want to know the total score of the Indian cricket team in a given match. To do so, your task is to find the sum of all the scores of the Indian team players. The scores are provided as a list, with each element as an individual score of the players. Also, there is a condition that if the number of elements in the list is more than 11, then it is an invalid input and the output should be -1.

--------------------------------------------------------------------------------------------
Input - List

Output - An integer

-------------------------------------------------------------------------------------------
Sample Input : [11, 13, 101, 14, 33, 141]

Sample Output : 313

--------------------------------------------------------------------------------------------
Sample Input : [11, 13, 101, 14, 33, 141, 12, 144, 54, 67, 8, 11]

Sample Output : -1

--------------------------

In [3]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
runs = ast.literal_eval(input())

# As there is an condition as if the no of elements exceed 11 then sum will (-1). to fulfill the condition, we will use if-else,

if len(runs)<=11:
    total_score=sum(runs)
    print(total_score)
else:
    total_score=-1
    print(total_score)
    

[11, 13, 101, 14, 33, 141]
313


In [5]:
# Method 2 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
runs = ast.literal_eval(input())


# here we can create a function to get the same using same ogic.
# As there is an condition as if the no of elements exceed 11 then sum will (-1). to fulfill the condition, we will use if-else,

def Cal_total_score(runs):
    if len(runs)>11:
        total_score=-1
        return total_score
    else:
        return (sum(runs))

#Printing the result by using created function:
print(Cal_total_score(runs))

[11, 13, 101, 14, 33, 141, 12, 144, 54, 67, 8, 11]
-1


In [66]:
"""ast = input()
sum = 0
b =""
for i in range(0,len(ast)):
    if len(ast) < 12:
        sum = sum + ast[i]
    else:
        b = "Invalid"
if b == "Invalid":
    print("-1")
else:    
    print(sum)"""

'ast = input()\nsum = 0\nb =""\nfor i in range(0,len(ast)):\n    if len(ast) < 12:\n        sum = sum + ast[i]\n    else:\n        b = "Invalid"\nif b == "Invalid":\n    print("-1")\nelse:    \n    print(sum)'

# 49.

Description

Given a list of strings, write a program to find the number of strings whose length is greater than or equal to K, where K is a positive integer.

Input - List of strings and an integer

Output - Integer

--------------------------------------------------------------------------------
Sample Input :

[Mumbai, Hyderabad, Calicut, Chennai]

  9

Sample Output: 1

------------------------------------------------------------------------------
Sample Input :

[Datascience, Data Analyst, Programmer, Manager]

  8
  
Sample Output : 3


In [12]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
String_list = ast.literal_eval(input())

# taking input of K, a positive integer
K=int(input())
#Initialising a count variable that will track the of element in the list
count=0

#Looping through the list and check the length of each element. Increase the count of word by 1 if the length is greater than K

for word in String_list:
    if len(word)>=K:
        count=count+1
print(count)

['Datascience', 'Data Analyst', 'Programmer', 'Manager']
8
3


In [2]:
import ast
input_list=ast.literal_eval(input())
k=int(input())
list1=[]

for ele in input_list:
    if len(ele)>=k:
        list1.append(ele)
print(len(list1))

['Datascience', 'Data Analyst', 'Programmer', 'Manager']
8
3


# 50.
Increment list elements

Description

Given a list of strings, increment the value of the numeric strings by 'k’. 

Hint: The function isdigit() may be useful here.

---------------------------------------------------------------------------------------------
Input - A list in the first line and an integer in the second line

Output - A list

---------------------------------------------------------------------------------------------
Sample Input :

['Python', '123', 'Data']

  4

Sample Output : ['Python', '127', 'Data']

---------------------------------------------------------------------------------------------
Sample Input :

['upGrad', '1991', 'Mumbai']

  0

Sample Output : ['upGrad', '1991', 'Mumbai']

---------------------------------------------------------------------------------------------

Sample Input :

['Data Science', '100', '10']

  10

Sample Output : ['Data Science', '110', '20']

In [5]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
String_list = ast.literal_eval(input())

# taking input of K, a positive integer
k =int(input())

# Created a empty list to store new elemnts of the list from the String_list
new_list=[]

# Initialing loop through string_list to check weather there is numeric value by isdigit() method , if true it will increase by k and new value will be added to the new list along old string which were not numeric. 

for i in String_list:
    
    if i.isdigit():
        i=int(i) + k
    new_list.append(str(i))
                 
print(new_list)


['kolkata','delhi']
5
['kolkata', 'delhi']


In [3]:
# Method 2
# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
String_list = ast.literal_eval(input())

# taking input of K, a positive integer
k =int(input())

# dinfining a function that take arguments as list and integer ad check if digits are there it will increase by integer
def check_digit(lst,k):
    for i in range(len(lst)):
        if lst[i].isdigit():
            #Increamenting on testing on the loop , increament by k in isdigit() is true or else continue the loop
            lst[i]=str(int(lst[i]) + k)
        else:
            continue
    return lst

# Calling the function
new_lst = check_digit(String_list,k)

# printing the result
print(new_lst)



['kolkata', 'delhi','500','101']
71
['kolkata', 'delhi', '571', '172']


# 51.

Slicing a list

Description

Given a list of strings and an integer K, write a python code to print all the elements from the K th position till the end of the list.

Note: Assume that K (a positive integer) will always be less than or equal to the length of the list

Input - A list of strings in the first line and an integer in the second line of the input.

Output - A list

--------------------------------------------------------------------------------------------------------
Sample Input :

['Mumbai', 'Delhi', 'Australia', 'Nigeria', 'USA', 'London', 'Canada']

2

Sample Output : ['Delhi', 'Australia', 'Nigeria', 'USA', 'London', 'Canada']

--------------------------------------------------------------------------------------------------------

Sample Input :

['Chennai', 'Vizag', 'Austria', 'Germany', 'Japan']

3

Sample Output : ['Austria', 'Germany', 'Japan']

In [9]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
string_list = ast.literal_eval(input())

# taking input of K, a positive integer
k =int(input())

# to print from thr kth position upto end
print(string_list[(k-1):])

['Chennai', 'Vizag', 'Austria', 'Germany', 'Japan']
3
['Austria', 'Germany', 'Japan']


## <font color= blue> Practice Questions II </font>

# 52.

Substring with maximum uppercase characters

Description

Given a string, write a Python program to find the largest substring of uppercase characters and print the length of that substring. Check the sample inputs and outputs for a better understanding.

---------------------------------------------------------------------------------------------------
Input - String

Output - String

---------------------------------------------------------------------------------------------------
Sample Input - I lovE PRogrAMMING

Sample Output - 6

Explanation - AMMING is the largest substring with all characters in uppercase continuously 

-----------------------------------------------------------------------------------------------------
Sample Input - MuMbaI is in MAHArashTRA

Sample Output - 4

Explanation - MAHA is the largest substring with all characters in uppercase continuously.

---------------------------------------------------------------------------------------------------
Sample Input - India WOn the WOrLD CUP

Sample Output - 3

Explanation - CUP is the largest substring with all characters in uppercase continuously.

In [19]:
# Method 1 

#Taking an input string
given_string = input()

#When a character in string is in upper case 'count' will count number of continuous upper case character from index 
count=0

#lst will store the each continuous count in a list format
lst=[]

# looping through the string to check the continuous character in uppercase 
for i in given_string:
    if i.isupper():
        count= count+1
        # encountering the character in lowercase the count will again be zero and start counting the continuous upper case again
        lst.append(count)
    else:
        count=0
print(max(lst))
        
        

BIdhAn
2


# 53.
Sorting list of tuples

Description

A class of students attempt an exam in two parts: ‘Aptitude’ and ‘Physics'. The marks of all the students are stored as a list of tuples, and each student’s marks (in Aptitude and Physics) are stored in each tuple. Your task is to write a Python program to sort the list of tuples in decreasing order of the Physics scores of the students.

Note - Marks of both the subjects are ranged between 1-100, and no two students have scored the same marks in Physics.

---------------------------------------------------------------------------------------------------
Input - List of tuples

Output - List of tuples

---------------------------------------------------------------------------------------------------
Sample Input - [(45,77), (88,87), (67,98), (33,31)]

In (45,77), which is the first element in the list, 45 and 77 are the scores of a student in aptitude and physics respectively.

Sample Output - [ (67,98), (88,87), (45,77),(33,31)]

---------------------------------------------------------------------------------------------------
Sample Input - [(45,23), (45,88), (45,98), (45,44)]

Sample Output - [ (45,98), (45,88), (45,44),(45,23)]

In [32]:
# Method 1 

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
string_list = ast.literal_eval(input())

# Defining key of the sorted function. it is set to sort by the second item in the tuple
def rank(p):
    return p[1]

#printing the sorted list in decreasing order of key value
print(sorted(string_list, key=rank, reverse=True))


[(45,77), (88,87), (67,98), (33,31)]
[(67, 98), (88, 87), (45, 77), (33, 31)]
None


In [33]:
# Method 2

# To make the code more generic would like too take input from user and for that importing a module called 'ast' and here to take any type(here the input will a list) of data structure as input literal_eval() method is used.
import ast
string_list = ast.literal_eval(input())

#defining the fuction to sort in decreasing order by the 2nd item of a list of tuples
def sort_tuple_list(list_of_tuples):
    return sorted(list_of_tuples, key = rank, reverse = True)

# Defining key of the sorted function. it is set to sort by the second item in the tuple
def rank(p):
    return p[1]

#printing the sorted list in decreasing order of key value
print(sort_tuple_list(string_list))


[(45,77), (88,87), (67,98), (33,31)]
[(67, 98), (88, 87), (45, 77), (33, 31)]


# 54.
Divisibility check

Description

Given an integer ‘n’, your task is to write a Python code to find whether ‘n’ is divisible by all its digits or not. If they divide the number, then the number ‘n’ is a happy number. Otherwise, it is a sad number.

The number ‘n’ may be provided with commas. At first, you have to clean the number (by removing the commas involved) and then check whether the number is happy or sad.

---------------------------------------------------------------------------------------------------
Input - String

Output - String

---------------------------------------------------------------------------------------------------
Sample Input - 2,128

Sample Output - Happy Number

---------------------------------------------------------------------------------------------------
Sample Input - 256

Sample Output - Sad Number

---------------------------------------------------------------------------------------------------
Sample Input - 1124

Sample Output - Happy Number

In [75]:
# Method 1

n = input()

#If the n is provider with commas , to clean that replace() will going to be used. replacing commas with blank:
n = n.replace(',','')

#setting the flag as true .for that the section in the loop will run continuously until any kind of event makes it false. then the loop will stop
mark = True

for i in n:
    if int(i)!=0 and int(n)%int(i)==0:
        mark=True
    else:
        mark=False
        break
        
#if the input number haven't faced any kind event that makes the flag false throughout the loop will print "Happy Number" and else "Sad Number"
if mark:
    print("Happy Number")
else:
    print('Sad Number')
    
    

256
Sad Number


## <font color= blue> Practice Questions III </font>

# 55.
Updating the dictionary keys

Description

You are provided with a dictionary containing the names of different football clubs as keys and the name of the main player of the corresponding team as values. When the main player of a team retires, one of their teammates steps up to fill in their role as the main player.

You’re also provided with a list which contains the names of the football clubs for which the current main players are retiring and the names of the corresponding new main player. Your task is to update the values in the original dictionary with the names of the new main players.

------------------------------------------------------------------------------------------------------
Input - The input consists of a dictionary with different football clubs' names and their main player. In the next line, you will be provided with a list of the new players who have taken up the main roles.

Output - Dictionary

---------------------------------------------------------------------------------------------------
Sample Input :

{Barcelona:’Messi’ , ‘Real Madrid’: ‘Benzema’, ‘PSG’: ‘Neymar’ }

[[‘Barcelona’, ‘Griezmann’], [‘PSG’, ‘Ramos’]]

Sample Output: {Barcelona:‘Griezmann’, ‘Real Madrid’: ‘Benzema’, ‘PSG’: ‘Ramos’}

---------------------------------------------------------------------------------------------------
Sample Input:

{Liverpool:’A’, ‘Real Madrid’: ‘B’, ‘Chelsea’: ‘C’ }

[[Liverpool, ‘R’]]

Sample Output: {Liverpool:’R’, ‘Real Madrid’: ‘B’, ‘Chelsea’: ‘C’ }

In [14]:
# Method 1 

# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
#Taking input present club and player as keys and values of the dictionary respectively
dict1 = ast.literal_eval(input())

#taking input of new star payers list containing team and player respectively 
list1 = ast.literal_eval(input())

# Using Dictionary comprehenion list1 converted to dictionary
dict2={x[0]:x[1] for x in list1 }

# the value of input dictionary has been updated by using update() method 
dict1.update(dict2)

# printing the the input dictionary after updation :
print(dict1)

{'Barcelona':'Messi' , 'Real Madrid': 'Benzema', 'PSG': 'Neymar' }
[['Barcelona', 'Griezmann'], ['PSG', 'Ramos']]
{'Barcelona': 'Griezmann', 'Real Madrid': 'Benzema', 'PSG': 'Ramos'}


In [15]:
# Method 2
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
#Taking input present club and player as keys and values of the dictionary respectively
dict1 = ast.literal_eval(input())

#taking input of new star payers list containing team and player respectively 
list1 = ast.literal_eval(input())

# Using type conversion using list items,list1 converted to dictionary
dict2=dict(list1)
# the value of input dictionary has been updated by using update() method 
dict1.update(dict2)

# printing the the input dictionary after updation :
print(dict1)

{'Barcelona':'Messi' , 'Real Madrid': 'Benzema', 'PSG': 'Neymar' }
[['Barcelona', 'Griezmann'], ['PSG', 'Ramos']]
{'Barcelona': 'Griezmann', 'Real Madrid': 'Benzema', 'PSG': 'Ramos'}


In [30]:
# Method 3
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
#Taking input present club and player as keys and values of the dictionary respectively
dict1 = ast.literal_eval(input())

#taking input of new star payers list containing team and player respectively 
list1 = ast.literal_eval(input())

# looping through the input dictionary and input list updating the key and value from the list.

for x,y in dict1.items():
    # dict1.items() return a key,value pair that represents x,y
    for item in list1:
        if item[0]== x:
            dict1[x]=item[1]
            # compairing if dictionary key is equal to first item in the nested list , it it is match then the dict value will be the next item of the nested list
        else:
            pass

# printing the updated dictionary
print(dict1)
            

{'Barcelona':'Messi' , 'Real Madrid': 'Benzema', 'PSG': 'Neymar' }
[['Barcelona', 'Griezmann'], ['PSG', 'Ramos']]
{'Barcelona': 'Griezmann', 'Real Madrid': 'Benzema', 'PSG': 'Ramos'}


In [32]:
# Method 4

import ast
input_dict = ast.literal_eval(input())
input_list = ast.literal_eval(input())

def update_dict(d,l):
    for i in d.keys():
        for j in l:
            if j[0] == i:
                d[i] = j[1]
            else:
                pass
    return d

print(update_dict(input_dict,input_list))

{'Barcelona':'Messi' , 'Real Madrid': 'Benzema', 'PSG': 'Neymar' }
[['Barcelona', 'Griezmann'], ['PSG', 'Ramos']]
{'Barcelona': 'Griezmann', 'Real Madrid': 'Benzema', 'PSG': 'Ramos'}


# 56.

List average

Description

Suppose you are working as a Marketing Lead in a company and you want to recruit new employees in your team (which already consists of 10 members) to carry out the company’s major project. You select the employees based on five parameters: A, B, C, D and E. The value of each parameter is represented on a scale of 1 to 10. The present average value of these parameters for your team is given to you in a list (in the same order). You recruit a person only if their scores do not reduce the average score of more than two parameters of your team. If they are recruited, print the output as ‘Selected’. If not, print the output as ‘Rejected’.

--------------------------------------------------------------------------------------------------
Input - Two lists, where the first list is the average scores of the team in the five parameters, and the second list is the scores of the new employee in all the five parameters. The parameter order is the same for both the lists, which is A, B, C, D and E.

Output - String

--------------------------------------------------------------------------------------------------
Sample Input - [‘8’, ‘4’, ‘6’, ‘9’, ‘7’]

  [‘1’, ‘1.1’, ‘1.2’, ‘1.2’, ‘2.3’]

Sample Output - Rejected

--------------------------------------------------------------------------------------------------
Sample Input - [‘10’, ‘5’, ‘6’, ‘9’, ‘7’]

  [‘10’, ‘ 9.8’, ‘7.2’, ‘1.66’, ‘4.3’]

Sample Output - Selected

--------------------------------------------------------------------------------------------------
Sample Input - [‘8’, ‘5.66’, ‘6.5’, ‘10’, ‘10’]

  [‘7’, ‘ 10’, ‘6’, ‘7’, ‘9’]

Sample Output - Rejected

In [45]:
# Method 1

# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast
#Taking input Represents average scores of team of the recruiter manager in category A,B,C,D,E
list1 = ast.literal_eval(input())

#taking input of result scored by the employee in category A,B,C,D,E
list2 = ast.literal_eval(input())

# count variable is created to count the num of time for the condition is true
count=0

# looping through the 
for i in range(len(list1)):
    if float(list1[i])>=float(list2[i]):
        count+=1
if count<=2:
    print("Rejected")
else:
    print("Selected")


['11','10','6','7','9']
['8','7','8','7','10']
Selected


# 57.

Merge dictionaries

Description

Write a python code to merge two dictionaries into a single dictionary.

--------------------------------------------------------------------------------------------------
Input: Two dictionaries, one in each line

Output: A Dictionary
    
--------------------------------------------------------------------------------------------------
Sample Input :

{'a': 10, 'b': 8}

{'d': 6, 'c': 4}

Sample Output : {'c': 4, 'a': 10, 'b': 8, 'd': 6}

--------------------------------------------------------------------------------------------------
Sample Input :

{'a': 110, 'b': 88}

{'d': 62, 'c': 44}

Sample Output : {'a': 110, 'b': 88, 'd': 62, 'c': 44}

In [48]:
# Method 1
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

# using update() method to merge dictionery and printing the merged dictionary
dict1.update(dict2)
             
print(dict1)


{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


In [54]:
# Method 2
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

# Creating a user defined function by using updat() method to update input dictionary1.

def merge_dict(dict1,dict2):
    dict1.update(dict2)
    return dict1

#Printing result:

print(merge_dict(dict1,dict2))

{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


In [55]:
# Method 3
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

#Merging to dictionary using **method by creating a new dictionary without affecting the input dictionaries
dict3={**dict1,**dict2}
print(dict3)

{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


In [49]:
# Method 4
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

# Creating a user defining function where **method has been used to create a merged dictionary without affecting any input dict.
def merge_dict(dict1,dict2):
    dict3={**dict1,**dict2}
    return dict3

# print the merged dictionary
print(merge_dict(dict1,dict2))

{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


In [56]:
# Method 5
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

# using '|' operator only applicabale in python 3.9 and above. it creates a new merged dictionary 
dict3 = dict1 | dict2

# printing the dictionary
print(dict3)
    


{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


In [58]:
# Method 6
# taking input from user and for that importing a module called 'ast' and here to take any type(here the input will a dictionary) of data structure as input literal_eval() method is used.
import ast

#Taking two user input dictionary
dict1 = ast.literal_eval(input())

dict2 = ast.literal_eval(input())

# Creating a user defining function where '|' operation is used to create a merged dictionary without affecting any input dict.
def merge_dict(dict1,dict2):
    dict3 = dict1 | dict2
    return dict3

# print the merged dictionary
print(merge_dict(dict1,dict2))

{'x': 11, 'y': 28}
{'p':29,'q': 222}
{'x': 11, 'y': 28, 'p': 29, 'q': 222}


# <font color=BLUE>B. Control Structures and Functions in Python</font>

### <font color=BLUE>If-else, If-elif-else, for loop, while loop, lambda,map, filter, reduce</font>

In [61]:
# 1.
What is the output of the following code?

X = 12

if (X > 10 & X < 15):
print('YES')
else:
print('No')

IndentationError: expected an indented block (3877053997.py, line 7)

In [63]:
# 2.
#Complete the piece of code to print a maximum of three variables, a, b and c.

a=10
b=16
c=20

if(a > b and a > c):
    print("a")
elif( b > a and b > c):
    print("b")
else:
     print("c")

c


In [71]:
# 3.
#Which of the following sequences of inputs would give the output 'C' 'A' 'D' 'B' over multiple iterations?

#Given the code below, the output obtained in several runs is 'C' 'A' 'D' 'B'.
score =70
if score >= 90:
    print('A')
elif score >=80:
    print('B')
elif score >= 70:
    print('C')
elif score >= 60:
    print('D')
else:
    print('F')
    
#_____________

score =91
if score >= 90:
    print('A')
elif score >=80:
    print('B')
elif score >= 70:
    print('C')
elif score >= 60:
    print('D')
else:
    print('F')

#_____________

score =67
if score >= 90:
    print('A')
elif score >=80:
    print('B')
elif score >= 70:
    print('C')
elif score >= 60:
    print('D')
else:
    print('F')
    
#_____________

score =88
if score >= 90:
    print('A')
elif score >=80:
    print('B')
elif score >= 70:
    print('C')
elif score >= 60:
    print('D')
else:
    print('F')

C
A
D
B


In [72]:
# 4. What will the following segment of code print? Try doing this verbally.

if (10 < 0) and (0 < -10):
    print("A")
elif (10 > 0) or False:
    print("B")
else:
    print("C")

B


In [73]:
# 5. 
#What will the following segment of code print? Try solving it verbally.

if True or True:
    if False and True or False:
        print('A')
    elif False and False or True and True:
        print('B')
    else:
        print('C')
else:
    print('D')

B


# 6. 

if-Else

Description

Write a code to check if the string in input_str starts with a vowel or not. Print capital YES or NO.

For example, if input_str = 'analytics' then, your output should print 'YES'.

Sample Input:

alpha

Sample Output:

YES

In [7]:
# Method 1
# taking input from the user in lower case to avoid case sensitivity
str1 = input().lower()

# creating a user defined function to check whether the string starts with vowels
def starts_with_vowel(str1):
    if str1[0] in 'aeiou':
        print('YES')
    else:
        print("NO")

starts_with_vowel(str1)


analytics
YES


In [9]:
# Method 2
# taking input from the user in lower case to avoid case sensitivity
str1 = input().lower()

# using if-else to check whether the string starts with vowels 

if str1[0] in 'aeiou':
    print('YES')
else:
    print("NO")




Bidhan
NO


In [12]:
# 7.
#Complete this piece of code for calculating the element-wise difference between two lists and storing them in L3.

#[Assume that the lists are of the same length.]

L1 = [10, 20, 30, 24, 18]
L2 = [8, 14, 15, 20, 10]
L3=[]
for  i in range(len(L2)):
    L3.append(L2[i] -L1[i])
L3

[-2, -6, -15, -4, -8]

# 8.

Iterations

Description

You are given a list of string elements and asked to return a list which contains each element of the string in title case or in other words first character of the string would be in upper case and remaining all characters in lower case

Sample Input:
['VARMA', 'raj', 'Gupta', 'SaNdeeP']

Sample Output
['Varma', 'Raj', 'Gupta', 'Sandeep']

In [14]:
# Method 1 using title()
#Importing ast module
import ast

# To take inputs as a list , weusing literal_eval() function
input_list = ast.literal_eval(input())

# To store the results creating a new list
lst1=[]

#Initialising the loop through input_list to convert the all the elements to title case using title() and then stored in lst1
for i in input_list:
    lst1.append(i.title())
print(lst1)

['VARMA', 'raj', 'Gupta', 'SaNdeeP']
['Varma', 'Raj', 'Gupta', 'Sandeep']


In [16]:
# Method 2 using capitalize()
#Importing ast module
import ast

# To take inputs as a list , weusing literal_eval() function
input_list = ast.literal_eval(input())

# To store the results creating a new list
updated_list= []

#Initialising the loop through input_list to convert the all the elements to title case using capitalize() and then stored in updated_list
for i in input_list:
    updated_list.append(i.capitalize())
    
print(updated_list)


['VARMA', 'raj', 'Gupta', 'SaNdeeP']
['Varma', 'Raj', 'Gupta', 'Sandeep']


In [19]:
# 9.
#What's with all the Negativity?
#Which of the following would create a list of negative even numbers from -100 to 0 (-100 included and 0 not included)? [More than one option may be correct.]

print(list(range(0, -101, 2)))
print(list(range(-100, -1, 2)))
print(sorted(set(range(-2, -101, -2))))
l = []
for i in range(-100, 0):
    if(i % 2 == 0):
        l.append(i)
print(l)


[]
[-100, -98, -96, -94, -92, -90, -88, -86, -84, -82, -80, -78, -76, -74, -72, -70, -68, -66, -64, -62, -60, -58, -56, -54, -52, -50, -48, -46, -44, -42, -40, -38, -36, -34, -32, -30, -28, -26, -24, -22, -20, -18, -16, -14, -12, -10, -8, -6, -4, -2]
[-100, -98, -96, -94, -92, -90, -88, -86, -84, -82, -80, -78, -76, -74, -72, -70, -68, -66, -64, -62, -60, -58, -56, -54, -52, -50, -48, -46, -44, -42, -40, -38, -36, -34, -32, -30, -28, -26, -24, -22, -20, -18, -16, -14, -12, -10, -8, -6, -4, -2]
[-100, -98, -96, -94, -92, -90, -88, -86, -84, -82, -80, -78, -76, -74, -72, -70, -68, -66, -64, -62, -60, -58, -56, -54, -52, -50, -48, -46, -44, -42, -40, -38, -36, -34, -32, -30, -28, -26, -24, -22, -20, -18, -16, -14, -12, -10, -8, -6, -4, -2]


In [20]:
# 10.
#Using the following document, answer the following questions. What will the output of the following code be?
 
d = {0: 'Fish', 1: 'Bird', 2: 'Mammal'}
for i in d:
    print(i)

0
1
2


# 11.
List Comprehension

Description

You are given an integer 'n' as the input. Create a list comprehension containing the squares of the integers from 1 till n^2 (including 1 and n), and print the list. 

For example, if the input is 4, the output should be a list as follows:

[1, 4, 9, 16]

The input integer 'n' is stored in the variable 'n'. 

In [23]:
#Method 1 By using list comprehension
# taking an integer as input
n=int(input())

result=[x**2 for x in range(0,(n+1))]
print(result)

5
[0, 1, 4, 9, 16, 25]


In [24]:
# Method 2 By using loop
n=int(input())

list2=[]
for x in range(n+1):
    x=x**2
    list2.append(x)
print(list2)


5
[0, 1, 4, 9, 16, 25]


# 12.
List Comprehensions

Description

Extract the words that start with a vowel from a list input_list=[wood, old, apple, big, item, euphoria] using list comprehensions.

Sample Input:

['wood','old','apple','big','item','euphoria']

Sample Output:

['old', 'apple', 'item', 'euphoria']

In [27]:
# Method 1 by list comprehension
import ast
input_list = ast.literal_eval(input())

output=[word for word in input_list if word[0].lower() in 'aeiou']

print(output)

['Apple','mango','English']
['Apple', 'English']


In [26]:
# Method 2 by using loop

import ast 
input_list= ast.literal_eval(input())
new_list = []
for word in input_list:
    if word[0] in 'aeiouAEIOU':
        new_list.append(word)
        
print(new_list)

['wood','old','apple','big','item','euphoria']
['old', 'apple', 'item', 'euphoria']


In [38]:
# 13.
#Following is the code to create a dictionary where the keys are multiples of 3 among the first 100 natural numbers and each value is the cube of the key.

input_list = list(range(1,100))
output_dict = {}

for val in input_list:
    if val % 3 == 0:
        output_dict[val] = val**3
print('1. ',output_dict)
#Now, what would be corresponding dictionary comprehension for the code involved in dictionary creation? 
print('anwer(2-5) which will match with 1 is the correct ans')

1.  {3: 27, 6: 216, 9: 729, 12: 1728, 15: 3375, 18: 5832, 21: 9261, 24: 13824, 27: 19683, 30: 27000, 33: 35937, 36: 46656, 39: 59319, 42: 74088, 45: 91125, 48: 110592, 51: 132651, 54: 157464, 57: 185193, 60: 216000, 63: 250047, 66: 287496, 69: 328509, 72: 373248, 75: 421875, 78: 474552, 81: 531441, 84: 592704, 87: 658503, 90: 729000, 93: 804357, 96: 884736, 99: 970299}
anwer(2-5) which will match with 1 is the correct ans


In [34]:
print('2. ',([val:val**3 for val in input_list if val%3=0]))

SyntaxError: invalid syntax (1565504096.py, line 1)

In [35]:
print('3. ',({for val in input_list if val%3==0 val:val**3 }))

SyntaxError: invalid syntax (3949294981.py, line 1)

In [36]:
print('4. ',({if val%3==0 val:val**3 for val in input_list}))

SyntaxError: invalid syntax (2399673509.py, line 1)

In [37]:
print('5. ',({val : val**3 for val in input_list if val % 3 == 0}))

5.  {3: 27, 6: 216, 9: 729, 12: 1728, 15: 3375, 18: 5832, 21: 9261, 24: 13824, 27: 19683, 30: 27000, 33: 35937, 36: 46656, 39: 59319, 42: 74088, 45: 91125, 48: 110592, 51: 132651, 54: 157464, 57: 185193, 60: 216000, 63: 250047, 66: 287496, 69: 328509, 72: 373248, 75: 421875, 78: 474552, 81: 531441, 84: 592704, 87: 658503, 90: 729000, 93: 804357, 96: 884736, 99: 970299}


In [40]:
#14.
#What will the output of the following code be?

print([i+j for i in "abc" for j in "def"])

['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']


In [41]:
# 15.
#What will the output of the following code be?

d = {x.upper(): x*3 for x in 'acbd'}
print(d)

{'A': 'aaa', 'C': 'ccc', 'B': 'bbb', 'D': 'ddd'}


# 16.
Function

Description

Create a function squared(), which takes x and y as arguments and returns the x**y value. For e.g., if x = 2 and y = 3 , then the output is 8.

Sample Input:

['6','7']

Sample Output:

279936

In [44]:
# Method 1
import ast

arg1= int(input())
arg2= int(input())

def squared(arg1,arg2):
    result= arg1**arg2
    return result

print(squared(arg1,arg2))

4
4
256


In [46]:
# 17. 
#Identify the error/errors in the given code to apply the function (x∗ y+x+y) on x,y and finally return the value.

'''
func(x, y):
    z = x * y + x + y

func(2)
'''

print('''
ans:
a. Keyword missing in the function definition
b. Return the statement missing in the code
c. Error in function call
''')


ans:
a. Keyword missing in the function definition
b. Return the statement missing in the code
c. Error in function call



In [47]:
# 18.
#What would be the output of the following code?

def func(x, y = 1):
    z = x * y + x + y
    return z

func(2, func(3))

23

In [48]:
# 19.
#What would be the output of the following code?

def func(x, y):
    z = x * y + x + y
    return z

func(2, func(3))

TypeError: func() missing 1 required positional argument: 'y'

In [49]:
# 20.
#What is the output of the following code?

def func(x = 1 ,y = 2):
    z = x * y + x + y
    return z

func(2, func(3))

35

In [50]:
# 21.
#Functions can be used inside a list comprehension too. An example to create a list, the elements of which would be the result of applying the function x2−2x−2on each element in the list = [2, 7, 8, 10, 3] is shown below:

L1 = [2, 7, 8, 10, 3]

def func(y):
    return y**2 - 2 * y - 2

ans_list = [func(x) for  x in L1]
ans_list

[-2, 33, 46, 78, 1]

# 22.

Lambda

Description

Create a lambda function 'greater', which takes two arguments x and y and return x if x>y otherwise y.

If x = 2 and y= 3, then the output should be 3.

Sample Input:

['9','3']

Sample Output:

9

In [59]:
# Method 1 by Lambda function
import ast

arg1 = int(input())
arg2 = int(input())

greater = lambda arg1,arg2: arg1 if arg1>arg2 else arg2

print(greater(arg1,arg2))

51
101
101


In [60]:
# Method 2 by if-else

import ast

x = int(input())
y = int(input())

if x>y:
    print(x)
else:
    print(y)

5
6
6


In [61]:
# 23.
#What is the output of this program?

min = (lambda x, y: x if x < y else y)
min(101*99, 102*98)

9996

In [68]:
# 24.
#A word is said to belong to the red team if it has the letter r in it. Write an 'if' statement to check whether the given word belongs to the red team or not.

# write code here
str1= input()

if 'r' in str1:
    print(f'the word {str1} belongs to red team')
else:
    print(f'the word {str1} doesn\'t belongs to red team')  

wgite
the word wgite doesn't belongs to red team


# 25.
In continuation to previous example, if the word has the letter 'b', it belongs to the blue team. And if the word has both 'r' and 'b' then first letter gets a precedence. And if the word does not have the letter 'r' or 'b', it does not belong to any team. ######Example: (word,output)

rabbit = red team

brand = blue team

dog = no team

Write a condition for this decision.

In [83]:
word = input("Enter a word in small caps ")
is_red = 0
is_blue = 0
for char in word:
    if char == "r":
        print ("The word belongs to red team")
        is_red = 1
        break
    elif char == "b": 
        print ("The word belongs to blue team")
        is_blue = 1
        break
        
if is_red == 0 and is_blue == 0:
    print ("The word does not belong to any team")

Enter a word in small caps blue
The word belongs to blue team


In [86]:
# 26.
#Question
#Write a program to print all the numbers which are divisible by 17 between 1 and a given number 'n'.
#Enter the upper limit 100

n = int(input('Enter the upper limit:'))
for i in range(1,n):
    if i%17==0:
        print(i)

Enter the upper limit:100
17
34
51
68
85


In [101]:
# 27.
#Write a program to add 13 to all the elements of a given numeric list. ######Example -

#Input list = [1,5,27]
#Output list = [14,18,40]

import ast
list1= [1,5,27]

list2=[]

for i in list1:
    list2.append(i+13)
        
print(list2)

[14, 18, 40]


In [103]:
# 28.
#In continuation to the example above, execute the same operation on the same sample list using list comprihension.

list2 = [x+13 for x in list1]
print(list2)

[14, 18, 40]


# 29.
Question
This question has two tasks:

Task 1 - Find the factors of all the numbers in a given list.
Task 2 - Add the numbers and their factors to a dictionary where the key will be the numbers from the list, and the values will be the factors of the respective number.
Example -

Sample output :
#####{6: [1, 2, 3, 6], 98: [1, 2, 7, 14, 49, 98],

34: [1, 2, 17, 34],}

In [113]:
list1 = [6, 98, 34, 45, 64, 7, 2, 5, 78, 90, 324, 111, 657, 438]

dict1={}
for i in list1:
    fact =[]
    for a in range(1,(i+1)):
        if i%a==0:
            fact.append(a)
    dict1[a] = fact
print(dict1)

{6: [1, 2, 3, 6], 98: [1, 2, 7, 14, 49, 98], 34: [1, 2, 17, 34], 45: [1, 3, 5, 9, 15, 45], 64: [1, 2, 4, 8, 16, 32, 64], 7: [1, 7], 2: [1, 2], 5: [1, 5], 78: [1, 2, 3, 6, 13, 26, 39, 78], 90: [1, 2, 3, 5, 6, 9, 10, 15, 18, 30, 45, 90], 324: [1, 2, 3, 4, 6, 9, 12, 18, 27, 36, 54, 81, 108, 162, 324], 111: [1, 3, 37, 111], 657: [1, 3, 9, 73, 219, 657], 438: [1, 2, 3, 6, 73, 146, 219, 438]}


# 30.
Count the total number of unique characters in a given sentence. #####Example -

Input - 'total'

Output - 'o', 'a', 'l'

In [119]:
x=input()
unique=set(x)
print(unique)
print(len(unique))

i am bidhan chandra roy
{'y', 'm', 'o', 'c', 'r', ' ', 'b', 'a', 'n', 'i', 'h', 'd'}
12


# 31.
Write a calculator application using functions, which will ask for the users input such as - arithmetic operation, number1, number 2, and return the output post applying the arithmetic operator

#####Example -

def calc(operation, x,y):
    if operation == "add":
        return x+y

In [3]:
def calc(operation, x,y):
    if operation == "add":
        return x+y
    if operation == "subs":
        return x-y
    if operation == "multi":
        return x*y
    if operation == "div":
        return x/y
    else:
        print("Enter a valid operation")
        return None
    
print ("Enter the operation you want to execute")
oper = input("The available operations are 'add', 'subs', 'multi', 'div' " )
x = int (input ("Enter the first number :"))
y = int (input ("Enter the second number :"))

print("Answer is ", calc(oper, x, y))

Enter the operation you want to execute
The available operations are 'add', 'subs', 'multi', 'div' subs
Enter the first number :25
Enter the second number :11
Answer is  14


In [11]:
# 32.
#Question
#Write a code to ask the user to input three numbers and then print the largest one among the three.

a=int(input())
b=int(input())
c=int(input())

if a>b and a>c:
    print(a)
elif b>a and b>c:
    print(b)
else:
    print(c)

4
6
3
6


# 33.
Check whether a number is perfect or not.

A perfect number is a number that is sum of all of its positive divisors (excluding itself). For more info refer to this link - https://www.geeksforgeeks.org/perfect-number/

In [13]:
n = int(input())

sum=0
for i in range(0,n):
    if n%2 == 0:
        sum=sum+i
        
if sum==n:
    print('perfect number')
else:
    print('not a perfect number')

28
not a perfect number


In [21]:
# 34.
# Add the elements in the two given lists using map and lambda functions.
list1 = [5, 8, 9, 12]
list2 = [10, 12, 8, 6]
list3=list(map(lambda x,y: x+y,list1,list2))
print(list3)

[15, 20, 17, 18]


In [22]:
# 35.
#Write a code to find the factorial of a number using reduce function. #####Example - 5!= 5 * 4 * 3 * 2 * 1=120.
n = int(input())
n_list= range(1,n+1)
from functools import reduce
factorial = reduce(lambda x,y:x*y,n_list)
print(factorial)

5
120


# 36.
Map Function

Description

Using the function Map, count the number of words that start with ‘S’ in input_list.

Sample Input:

['Santa Cruz','Santa fe','Mumbai','Delhi']


In [28]:
# Method 1 by using map and lambda function

import ast

input_list = ast.literal_eval(input())

check = list(map(lambda x: x[0] in 'S',input_list))

output = check.count(True)

print(output)

['Santa Cruz','Santa fe','Mumbai','Delhi']
2


In [30]:
# Method 2

import ast

input_list = ast.literal_eval(input())

count=0

for x in input_list:
    if x.startswith('S'):
        count+=1
print(count)
    

['Santa Cruz','Santa fe','Mumbai','Delhi']
2


In [32]:
# Method 3

import ast

input_list = ast.literal_eval(input())

count=0

for x in input_list:
    if x[0] in 'S':
        count+=1
print(count)

['Santa Cruz','Santa fe','Mumbai','Delhi']
2


# 37.

Map Function

Description

Create a list ‘name’ consisting of the combination of the first name and the second name from list 1 and 2 respectively. 

For e.g. if the input list is:
[ ['Ankur', 'Avik', 'Kiran', 'Nitin'], ['Narang', 'Sarkar', 'R', 'Sareen']]

the output list should be the list:
['Ankur Narang', 'Avik Sarkar', 'Kiran R', 'Nitin Sareen']

Note: Add a space between first name and last name.

In [36]:
import ast

input_list = ast.literal_eval(input())

first_name=input_list[0]
second_name=input_list[1]

name=list(map(lambda x,y: x+' '+y,first_name,second_name))
print(name)

[ ['Ankur', 'Avik', 'Kiran', 'Nitin'], ['Narang', 'Sarkar', 'R', 'Sareen']]
['Ankur Narang', 'Avik Sarkar', 'Kiran R', 'Nitin Sareen']


# 38.
Filter Function

Description

Extract a list of numbers that are multiples of 5 from a list of integers named input_list.

Sample Input:

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]

Sample Output:

[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]

Note: Use the filter() function.

In [39]:
import ast

input_list = ast.literal_eval(input())

output = list(filter(lambda x:x%5==0,input_list))
output

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]

# 39.

Filter Function

Description

You are given a list of strings such as input_list = ['hdjk', 'salsap', 'sherpa'].

Extract a list of names that start with an ‘s’ and end with a ‘p’ (both 's' and 'p' are lowercase) in input_list.

Sample Input:

['soap','sharp','shy','silent','ship','summer','sheep']

Sample Output:

['soap', 'sharp', 'ship', 'sheep']

Note: Use the filter() function.

In [42]:
import ast

input_list = ast.literal_eval(input())

output=list(filter(lambda x: x[0]=='s' and x[-1]=='p',input_list))
output

['soap','sharp','shy','silent','ship','summer','sheep']


['soap', 'sharp', 'ship', 'sheep']

# 40.

Reduce Function

Description

Using the Reduce function, concatenate a list of words in input_list, and print the output as a string.

If input_list = ['I','Love','Python'], the output should be the string 'I Love Python'.

Sample Input:

['All','you','have','to','fear','is','fear','itself']

Sample Output:

All you have to fear is fear itself

In [44]:
import ast

input_list=ast.literal_eval(input())

from functools import reduce

output= reduce(lambda x,y: x+" "+y,input_list)
output

['All','you','have','to','fear','is','fear','itself']


'All you have to fear is fear itself'

# 41.

Reduce Function

Description

You are given a list of numbers such as input_list = [31, 63, 76, 89]. Find and print the largest number in input_list using the reduce() function.

Sample Input:

[65,76,87,23,12,90,99]

Sample Output:

99

In [46]:
import ast

input_list = ast.literal_eval(input())

max_num = reduce(lambda x,y: x if x>y else y, input_list )

max_num

[65,76,87,23,12,90,99]


99

## <font color= blue>(Week-3) Bootcamp Practice Questions </font>
### <font color= blue> Practice Questions MCQ </font>

In [60]:
# 42.
#What will be the output of the following code?

input_string = "I love Python"
k = [(i.upper(), len(i)) for i in input_string.split(' ')]
print(k)

[('I', 1), ('LOVE', 4), ('PYTHON', 6)]


In [61]:
# 43.
#What will be the output of the following code snippet?

var = [i**2 for i in range(5)]
print(var)

[0, 1, 4, 9, 16]


In [62]:
# 44.
#What will be the output of the following code?

l1=[1,2,3,7]
l2=[4,5,6,9]
[x+y for x in l1 for y in l2]

[5, 6, 7, 10, 6, 7, 8, 11, 7, 8, 9, 12, 11, 12, 13, 16]

In [63]:
# 45.
# Functions in Python
#What will be the output of the following?

def function(n):
  n=n+1
 
n=2
print(function(2))

None


In [64]:
# 46.
#What will be the output of the below code?

x=10
def func(x):
   x=20
print(x)
 
print(x)

10
10


In [65]:
# 47
#What will be the output of the following?

def samplefunc(msg,i=3):
  print(msg*i)
 
samplefunc('python',4)
samplefunc('program')

pythonpythonpythonpython
programprogramprogram


In [66]:
#48
#What will be the output of the following program?

f=lambda x:bool(x%2)
print(f(100)* f(101))

0


In [67]:
# 49.
#What will be the answer printed when the following code is executed?

l=(1, -2, -3, 4, 5)
def f1(x):
    return x//2
m1=map(f1, l)
print(list(m1))

[0, -1, -2, 2, 2]


In [68]:
#50.
#What will be the output of the code below?

from functools import reduce
numbers = [-1, 2, -5, 3, 8,-2, 1,-4,-2]
sum = reduce(lambda x, y: x + y, numbers)
print(sum)

0


In [69]:
# 51.
#What will be the output of the following code?

sample=[-4,2,5,-3]
ans=map(lambda x: x*2, sample)
print(ans)

In [70]:
# 52.
#What will be the output of the following code snippet?

from functools import reduce
print(list(reduce(lambda x,y:x+y,range(5,15))))

TypeError: 'int' object is not iterable

In [71]:
#53.
#Is the output of the following two code snippets the same? Are they logically the same for any values in ‘lis’?

import functools
lis=[1, 2, 3, 4, 5]
m=functools.reduce(lambda x, y:x if x<y else y, lis)
print(m)
lis=[1, 2, 3, 4, 5]
c=lis[0]
for i in range(len(lis)-1):
    if lis[i]<c:
       c=lis[i]
print(c)

1
1


In [72]:
#54
#What will be the output of the following code?

letters = ('d','i','a','s')
ans = list(map(lambda word: f"{word}x", letters))
print(ans)

['dx', 'ix', 'ax', 'sx']


In [73]:
#55.
#Read the following code. Select from the given options the length of the list printed. (Do not run it on the notebook. Try solving manually.)

def sf(a):
    return a%4==0 and a%8!=0
m=filter(sf, range(10, 31))
print(list(m))

[12, 20, 28]


## <font color= blue> Practice Questions I </font>

# 56.

Even numbers in a list

Description

Given a list of integers, write a python code to find all the even numbers present in the list.

Note: Try using lambda and filter functions to solve the problem.

Input: A list of integers

Output: A list of integers

----------------------------------------------------------------------
Sample input: [5, 7, 22, 97, 54, 62, 77, 23, 73, 61] 

Sample output: [22, 54, 62]

----------------------------------------------------------------------
Sample input: [500, 789, 2112, 9097, 5894, 9062, 8977, 1223, 773, 600]

Sample output: [500, 2112, 5894, 9062, 600]

In [82]:
#Method 1 using filter and lambda function

import ast

input_list = ast.literal_eval(input())

#since the result is a subset of the input list, we will use the filter function. 
#instead of for loop, we will use the lambda function.  

even_num = list(filter(lambda x: x%2==0,input_list))

print(even_num)

[500, 789, 2112, 9097, 5894, 9062, 8977, 1223, 773, 600]
[500, 2112, 5894, 9062, 600]


In [77]:
#Method 2 using loop

import ast

input_list=ast.literal_eval(input())

count=0
even_num = []

for x in input_list:
    if int(x)%2==0:
        count+=1
        even_num.append(x)
print(even_num)     


[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
[22, 54, 62]


In [79]:
# Method 3 using list comprehension

import ast

input_list=ast.literal_eval(input())

even_num = [x for x in input_list if (x)%2==0]

print(even_num)

[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
[22, 54, 62]


# 57.

Triple the list elements

Description

Given a list of integers, write a python code to triple all the values in the list.

Note: Try using lambda and map functions to solve the problem.

Input: A list of integers

Output: A list of integers

----------------------------------------------------------------------
Sample input: [5, 7, 22, 97, 54, 62, 77, 23, 73, 61]

Output: [15, 21, 66, 291, 162, 186, 231, 69, 219, 183]

---------------------------------------------------------------------
Sample input: [15, 71, 2, 9, 5, 6, 11, 3, 7, 6]

Output: [45, 213, 6, 27, 15, 18, 33, 9, 21, 18]

In [83]:
# Metheod 1 Using map and lambda function

import ast

input_list = ast.literal_eval(input())

#mapping lambda function to each element of the input_list

list_value_3x = list(map(lambda x: x*3, input_list))

print(list_value_3x)

[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
[15, 21, 66, 291, 162, 186, 231, 69, 219, 183]


In [84]:
# Method 2 using list comprehension

import ast

input_list = ast.literal_eval(input())

list_value_3x = [x*3 for x in input_list]

print(list_value_3x)


[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
[15, 21, 66, 291, 162, 186, 231, 69, 219, 183]


In [86]:
# Method 3 using loop

import ast

input_list = ast.literal_eval(input())

list_value_3x =[]

for x in input_list:
    x = x*3
    list_value_3x.append(x)

print(list_value_3x)

[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
[15, 21, 66, 291, 162, 186, 231, 69, 219, 183]


# 58.

Encrypt elements in a list

Description

A company stores the names of their employees in a list. They want to encrypt the names so that no one can read them and the data stays safe. One of the steps in encrypting is to reverse each name in the list and convert them to uppercase and your task is to write a Python code to execute the same.

Note: Use lambda and map functions.

Input: A list of names

Output: A list of encrypted names

----------------------------------------------------------------------
Sample input: ['Ronaldo', 'Cristiano', 'Rakesh', 'Ronak']

Sample output: ['ODLANOR', 'ONAITSIRC', 'HSEKAR', 'KANOR']

----------------------------------------------------------------------
Sample input: ['Abhay', 'Aravind', 'Desilva', 'Darpan']

Sample output: ['YAHBA', 'DNIVARA', 'AVLISED', 'NAPRAD']

In [89]:
# Metheod 1 Using map and lambda function

import ast

input_list = ast.literal_eval(input())

#Here, we intend to change all names by reversing the string and converting  them to upper case
encrypt_names =list(map(lambda x:x.upper()[::-1],input_list))

print(encrypt_names) 

['Ronaldo', 'Cristiano', 'Rakesh', 'Ronak']
['ODLANOR', 'ONAITSIRC', 'HSEKAR', 'KANOR']


In [90]:
# Method 2 using list comprehension

import ast

input_list = ast.literal_eval(input())

encrypt_names = [x.upper()[::-1] for x in input_list]

print(encrypt_names) 

['Abhay', 'Aravind', 'Desilva', 'Darpan']
['YAHBA', 'DNIVARA', 'AVLISED', 'NAPRAD']


In [91]:
# Method 3 using loop

import ast

input_list = ast.literal_eval(input())

encrypt_names =[]

for x in input_list:
    x=x.upper()[::-1]
    encrypt_names.append(x)
    
print(encrypt_names)

['Abhay', 'Aravind', 'Desilva', 'Darpan']
['YAHBA', 'DNIVARA', 'AVLISED', 'NAPRAD']


## <font color= blue> Practice Questions II </font>

# 59.

Sorting a dictionary

Description

Given is a list consisting of dictionary elements. Each dictionary contains the name of a student as a first item and his/her corresponding rank as the second item. Your task is to write a Python program to print the list (dictionary elements) in a sorted ascending order according to student ranks. In case the rank of two or more students are the same, then sort them in ascending order (alphabetically) according to their names.

Note: Use the sorted() function to solve the question.

Input: Dictionary

Output: Dictionary

----------------------------------------------------------------------
Sample input: [{ "name" : "Arpit", "rank" : 20}, { "name" : "Manjeet", "rank" : 20 },{ "name" : "Aravind" , "rank" : 19 }]

Sample output: [{ "name" : "Aravind" , "rank" : 19 }, { "name" : "Arpit", "rank" : 20}, { "name" : "Manjeet", "rank" : 20 },

]

----------------------------------------------------------------------
Sample input: [{ "name" : "Mohan", "rank" : 2}, { "name" : "Mandeep", "rank" : 2 },

{ "name" : "Merut" , "rank" : 2 }]

Sample output: [{'name': 'Mandeep', 'rank': 2}, {'name': 'Merut', 'rank': 2}, {'name': 'Mohan', 'rank': 2}]

In [99]:
# Metheod 1 Using sorted() and lambda function

import ast

input_list = ast.literal_eval(input())

sorted_list = sorted(input_list,key = (lambda D:(D['rank'], D['name'])))

print(sorted_list)

[{ "name" : "Mohan", "rank" : 2}, { "name" : "Mandeep", "rank" : 2 },  { "name" : "Merut" , "rank" : 2 }]
[{'name': 'Mandeep', 'rank': 2}, {'name': 'Merut', 'rank': 2}, {'name': 'Mohan', 'rank': 2}]


In [98]:
# Method 2 using itemgetter()

import ast
from operator import itemgetter

input_list = ast.literal_eval(input())

sorted_list = sorted(input_list,key = itemgetter('rank','name'))

print(sorted_list)

[{ "name" : "Arpit", "rank" : 20}, { "name" : "Manjeet", "rank" : 20 },{ "name" : "Aravind" , "rank" : 19 }]
[{'name': 'Aravind', 'rank': 19}, {'name': 'Arpit', 'rank': 20}, {'name': 'Manjeet', 'rank': 20}]


# 60.

Intersection of lists

Description

In ‘ABC’ university, company ‘XYZ’ conducted three rounds of interviews to recruit students. Those students who passed all the rounds were selected. The company released the results for each round in a separate list. Your task is to write a Python code to display the list of the students who passed all the three rounds.

Note: a. Name of the person can be given in uppercase or lowercase.

 b. Print the names in lowercase in the output.

Input: List

List

List

Output: List
    
----------------------------------------------------------------------
Sample input: ['Arkam', 'Bairstow', 'Cairy', 'Darpan'] 

 ['ARKAM', 'Bairstow', 'Cairy', 'Darpan', 'Dhoni', 'Sachin']

 ['arkam', 'bairstow', 'Cheteshwar', 'Dinesh']
    
Sample output: ['arkam', 'bairstow']

----------------------------------------------------------------------
Sample input: ['Pranay', 'Aditya', 'Deepesh', 'Sandesh']

['Pranoy', 'Surya', 'Cairy', 'Vignesh', 'ADITYA', 'SANDesh']

['Vardhan', 'Shailesh', 'aditya', 'sandesh']

Sample output: ['aditya', 'sandesh']

----------------------------------------------------------------------
Sample input: ['Siddharth', 'Gaurav', 'Prasad', 'Kuldeep', 'karna']

['siddharth', 'Gaurav', 'Prasad', 'Kuldeep', 'karna']

['Siddharth', 'gaurav', 'PRASAD', 'KuLDEEp', 'KARNA']

Sample output: ['siddharth', 'gaurav', 'prasad', 'kuldeep', 'karna']

In [109]:
# Metheod 1 Using map() , filter() and lambda funtion

import ast

input_list1 = ast.literal_eval(input())
input_list2 = ast.literal_eval(input())
input_list3 = ast.literal_eval(input())

#converting all the list of elements in lower case
input_list1 = list(map(lambda m:m.lower(),input_list1))
input_list2 = list(map(lambda m:m.lower(),input_list2))
input_list3 = list(map(lambda m:m.lower(),input_list3))

#Making a list of common elements in input_list1 & input_list2
list1_list2 = list(filter(lambda x: x in input_list1,input_list2))

#Making a list of common elements in input_list3 & list1_list2
list1_list2_list3 =list(filter(lambda x: x in input_list3,list1_list2))

#printing the list of common elements in all three list 
print(list1_list2_list3)


['Pranay', 'Aditya', 'Deepesh', 'Sandesh']
['Pranoy', 'Surya', 'Cairy', 'Vignesh', 'ADITYA', 'SANDesh']
['Vardhan', 'Shailesh', 'aditya', 'sandesh']
['aditya', 'sandesh']


In [111]:
# Metheod 1 Using map() and lambda funtion and loop

import ast

input_list1 = ast.literal_eval(input())
input_list2 = ast.literal_eval(input())
input_list3 = ast.literal_eval(input())

#converting all the list of elements in lower case
input_list1 = list(map(lambda m:m.lower(),input_list1))
input_list2 = list(map(lambda m:m.lower(),input_list2))
input_list3 = list(map(lambda m:m.lower(),input_list3))

#creating a blank list to store all the common elements
common_list=[]

# initialising loop to check whether the elements are present in all three lists
for item in input_list1:
    if item in input_list2 and input_list3:
        common_list.append(item)
        
print(common_list)

['Vardhan', 'Shailesh', 'aditya', 'sandesh']
['Pranoy', 'Surya', 'Cairy', 'Vignesh', 'ADITYA', 'SANDesh']
['Pranay', 'Aditya', 'Deepesh', 'Sandesh']
['aditya', 'sandesh']


# 61.

Sorting a list of dictionary elements

Description

A teacher decided to divide the class into groups to conduct fun activities. Each group had one or more members in it. The data is provided in the form of a list with dictionary elements. The key and value in each dictionary correspond to the student’s roll number and his/her name.

In case the length of two or more dictionaries are the same, then the dictionary that occurs first in the input list will also appear first among the other dictionaries of the same length.

Your task is to write a Python program to sort the list in descending order with respect to the group strength.

Note: Use lambda and sorted functions to solve the question.

Input: List

Output: List

----------------------------------------------------------------------
Sample input: [{4: 'Viking', 9: 'Happy', 10: 'Tony', 2: 'Rogers'}, {12: 'Vikram', 19: 'Vipan'}, {3: 'Kohli'}, {1: 'Ronaldo', 29: 'Harish', 7: 'Ganesh'}]

Sample output: [{4: 'Viking', 9: 'Happy', 10: 'Tony', 2: 'Rogers'}, {1: 'Ronaldo', 29: 'Harish', 7: 'Ganesh'}, {12: 'Vikram', 19: 'Vipan'}, {3: 'Kohli'}]

----------------------------------------------------------------------
Sample input: [{14: 'Darpan', 19: 'Harry', 10: 'Julia'}, {12: 'Ethan', 19: 'Rogers'}, {3: 'Kohli', 1: 'Neymar', 29: 'Sachin', 7: 'Ramesh'}]

Sample output: [{3: 'Kohli', 1: 'Neymar', 29: 'Sachin', 7: 'Ramesh'}, {14: 'Darpan', 19: 'Harry', 10: 'Julia'}, {12: 'Ethan', 19: 'Rogers'}]

In [118]:
# Metheod 1 Using sorted() and lambda funtion 

import ast

input_list = ast.literal_eval(input())

#performing sort using sorted(), lambda for filtering. 

sorted_list = sorted(input_list, key=(lambda D:len(D)), reverse=True)

print(sorted_list)

[{14: 'Darpan', 19: 'Harry', 10: 'Julia'}, {12: 'Ethan', 19: 'Rogers'}, {3: 'Kohli', 1: 'Neymar', 29: 'Sachin', 7: 'Ramesh'}]
[{3: 'Kohli', 1: 'Neymar', 29: 'Sachin', 7: 'Ramesh'}, {14: 'Darpan', 19: 'Harry', 10: 'Julia'}, {12: 'Ethan', 19: 'Rogers'}]
